File: 90-subsets-ii/90-subsets-ii.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> temp;
        sort(begin(nums), end(nums));
        fun(nums, ans, temp);
        return ans;
    }
private:
        void fun(vector<int> &nums,vector<vector<int>> &ans, vector<int> &arr, int i = 0) {
        if (i == nums.size()) {
            ans.push_back(arr);
            return;
        }
        // pick
        arr.push_back(nums[i]);
        fun(nums, ans, arr, i+1);
        arr.pop_back();
        // not - pick
        // remove duplicate
        while ((i+1) < nums.size() && nums[i] == nums[i+1]) i++;
        fun(nums, ans, arr, i+1);
    }
};

File: 90-subsets-ii/90-subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        tmp,res = [],[]
        nums.sort()
        def helper(i):
            if i >= len(nums):
                res.append(tmp[:])
                return
            tmp.append(nums[i])
            helper(i+1)
            tmp.pop()
            while i+1 < len(nums) and nums[i+1] == nums[i]:
                i+=1
            helper(i+1)
        helper(0)
        return res